packages/core/src/Actor.ts

```diff
@@ -1,59 +1,126 @@
 import {
   EventObject,
   Subscribable,
-  InvokeDefinition,
-  AnyEventObject
+  SCXML,
+  InvokeCallback,
+  InterpreterOptions,
+  ActorRef
 } from './types';
+import { MachineNode } from './MachineNode';
+import { Interpreter } from './interpreter';
+import {
+  Behavior,
+  startSignal,
+  ActorContext,
+  stopSignal,
+  createServiceBehavior,
+  createMachineBehavior,
+  createCallbackBehavior,
+  createPromiseBehavior,
+  createObservableBehavior
+} from './behavior';
+import { registry } from './registry';
+
+const nullSubscription = {
+  unsubscribe: () => void 0
+};
+
+export function isActorRef(item: any): item is ActorRef<any> {
+  return !!item && typeof item === 'object' && typeof item.send === 'function';
+}
+
+export function fromObservable<T extends EventObject>(
+  observable: Subscribable<T>,
+  parent: ActorRef<any>,
+  name: string
+): ActorRef<never> {
+  return new ObservableActorRef(
+    createObservableBehavior(observable, parent),
+    name
+  );
+}
+
+export function fromPromise<T>(
+  promise: PromiseLike<T>,
+  parent: ActorRef<any>,
+  name: string
+): ActorRef<never> {
+  return new ObservableActorRef(createPromiseBehavior(promise, parent), name);
+}
+
+export function fromCallback<TEvent extends EventObject>(
+  callback: InvokeCallback,
+  parent: ActorRef<any>,
+  name: string
+): ActorRef<SCXML.Event<TEvent>> {
+  return new ObservableActorRef(createCallbackBehavior(callback, parent), name);
+}
+
+export function fromMachine<TContext, TEvent extends EventObject>(
+  machine: MachineNode<TContext, any, TEvent>,
+  parent: ActorRef<any>,
+  name: string,
+  options?: Partial<InterpreterOptions>
+): ActorRef<TEvent> {
+  return new ObservableActorRef(
+    createMachineBehavior(machine, parent, options),
+    name
+  );
+}
+
+export function fromService<TContext, TEvent extends EventObject>(
+  service: Interpreter<TContext, any, TEvent>,
+  name: string = registry.bookId()
+): ActorRef<TEvent> {
+  return new ObservableActorRef(createServiceBehavior(service), name);
+}
+
+enum ProcessingStatus {
+  NotProcessing,
+  Processing
+}
+
+export class ObservableActorRef<TEvent extends EventObject, TEmitted>
+  implements ActorRef<TEvent, TEmitted> {
+  public current: TEmitted;
+  private context: ActorContext;
+  private behavior: Behavior<TEvent, TEmitted>;
+  private mailbox: TEvent[] = [];
+  private processingStatus: ProcessingStatus = ProcessingStatus.NotProcessing;
+  public name: string;
+
+  constructor(behavior: Behavior<TEvent, TEmitted>, name: string) {
+    this.behavior = behavior;
+    this.name = name;
+    this.context = {
+      self: this,
+      name: this.name
+    };
+    this.current = behavior.current;
```

Comment:
this is only set in the constructor - imho we should either:
1. rename it
2. remove it
3. update after setting a new behavior

Reply:
seems that maybe the intention was the 3rd option (?): https://github.com/davidkpiano/xstate/pull/1163/files#diff-aebc63c0de70d68a4924833fd0ea3a5fR50

Reply:
It shouldn't exist at all. We should move this out of the `Actor`/`ActorRef` and instead create an `ActorStream` (bikesheddable) that wraps the actor and caches the last emitted value to make available as `current`.

Reply:
k, definitely a potential for other PRs